packages/aws-cdk-lib/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -452,6 +492,53 @@ export class StateMachine extends StateMachineBase {
       }
     }
 
+    if (props?.kmsKey) {
+      props.kmsKey.addToResourcePolicy(new iam.PolicyStatement({
```

Comment:
Is this supposed to be on the execution IAM role or the KMS Key Policy?

The IAM role definitely needs permissions to call Decrypt and GenerateDataKey if encryption is enabled.

Additionally, the execution role needs permissions to use the KMS key for any encrypted activies.

Reply:
Did you test executions and logs for state machines and activities created with these L2 updates? Creating syntactically correct CloudFormation that deploys successfully is one thing, but that doesn't mean all the configurations enable the desired functionality.

Reply:
Previously I was just ensuring that the CFN template deployed successfully. With the next round of changes, I have been testing executions as well. I will also update the snapshot tests to start an execution and confirm if logs are generated to verify functionality.

Reply:
In the latest commit I verified my changes worked by provisioning the following and starting an execution. The logs were also successfully emitted in both cases (Lambda and Activity):

```
import * as cdk from "aws-cdk-lib";
import * as kms from 'aws-cdk-lib/aws-kms';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs'
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {
  StateMachine,
  Chain,
  DefinitionBody,
  Activity,
  StateMachineType,
  LogLevel,
} from "aws-cdk-lib/aws-stepfunctions";
import { Construct } from "constructs";
import path = require("path");

export class HelloCdkStack extends cdk.Stack {
  testStateMachine: cdk.aws_stepfunctions.StateMachine;
  testActivity: cdk.aws_stepfunctions.Activity;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const key = new kms.Key(this, 'Key used for encryption');
    const activityKmsKey = new kms.Key(this, 'ActivityKMSKey');
    const logGroupKey = new kms.Key(this, 'LogGroup Key');

    const activity = new Activity(this, 'TestActivity', {
      activityName: 'TestActivty',
      kmsKey: activityKmsKey,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(100),
    });

    const logGroup = new logs.LogGroup(this, 'MyLogGroup', {
      logGroupName: '/aws/vendedlogs/states/MyLogGroup',
      encryptionKey: logGroupKey,
    });
    logGroupKey.grantEncryptDecrypt(new iam.ServicePrincipal('logs.amazonaws.com'));

    key.addToResourcePolicy(new cdk.aws_iam.PolicyStatement({
      resources: ['*'],
      actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
      principals: [new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`)],
      conditions: {
        ArnEquals: {
          'kms:EncryptionContext:aws:logs:arn': cdk.Stack.of(this).formatArn({
            service: 'logs',
            resource: 'log-group',
            sep: ':',
            resourceName: logGroup.logGroupName,
          }),
        },
      },
    }));

    // Define the Lambda function resource
    const helloWorldFunction = new lambda.Function(this, 'HelloWorldFunction', {
      runtime: lambda.Runtime.NODEJS_20_X, // Choose any supported Node.js runtime
      code: lambda.Code.fromAsset('lib'), // Points to the lambda directory
      handler: 'hello.handler', //
    });

    const stateMachine = new StateMachine(this, 'StateMachineWithCMKWithCWLEncryption', {
      stateMachineName: 'StateMachineWithCMKWithCWLEncryption',
      // definitionBody: DefinitionBody.fromChainable(Chain.start(new tasks.StepFunctionsInvokeActivity(this, 'Activity', {activity: activity}))),
      definitionBody: DefinitionBody.fromChainable(Chain.start(new tasks.LambdaInvoke(this, 'HelloWorldLambda', {lambdaFunction: helloWorldFunction}))),
      stateMachineType: StateMachineType.STANDARD,
      kmsKey: key,
      kmsDataKeyReusePeriodSeconds: cdk.Duration.seconds(300),
      logs: {
        destination: logGroup,
        level: LogLevel.ALL,
        includeExecutionData: false,
      },
    });
  }
}
```

### Lambda code:
```
exports.handler = async (event) => {
    return {
        statusCode: 200,
        headers: { "Content-Type": "text/plain" },
        body: JSON.stringify({ message: "Hello, World!" }),
    };
};
```

### Activity Worker Code
```
// Load the SDK and UUID
var AWS = require("aws-sdk");
var uuid = require("uuid");
var stepfunctions = new AWS.StepFunctions( {region: 'us-east-1'});


AWS.config.getCredentials(function(err) {
    if (err) console.log(err.stack);
    // credentials not loaded
    else {
      console.log("Access key:", AWS.config.credentials.accessKeyId);
    }
  });

stepfunctions.getActivityTask({activityArn: '' }, (err, data) => {
    if (err){
        console.log(err);
    } else{
        console.log(data);
        var params = {
            output: JSON.stringify({
                "hello":"there"
            }),
            taskToken: data.taskToken
        }
        stepfunctions.sendTaskSuccess(params, function (err, data) {
            if (err) console.log(err, err.stack); // an error occurred
            else console.log(data);           // successful response
        });
    }
});
```

Reply:
I believe this policy is supposed to be on the log group key, not the state machine's key:
```typescript
 key.addToResourcePolicy(new cdk.aws_iam.PolicyStatement({
      resources: ['*'],
      actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
      principals: [new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`)],
      conditions: {
        ArnEquals: {
          'kms:EncryptionContext:aws:logs:arn': cdk.Stack.of(this).formatArn({
```

A Lambda Task is probably not necessary to validating state machine encryption (a Pass state would suffice)

Reply:
Fair, will just use a pass state. Yes you are right, I have update the code to add the policy on the log group key and not the state machine key.

Reply:
> I will update the code to add the policy on the log group key

You already granted encrypt and decrypt permissions with this line. Otherwise I don't think you'd be able to get the logs. It's a bit broader (no EncryptonContext and uses the global CWL service principal) but doesn't include `kms:Describe*` (perhaps not necessary)?

```
    logGroupKey.grantEncryptDecrypt(new iam.ServicePrincipal('logs.amazonaws.com'));
```

Reply:
Yeah I noticed that as well. If I remove that line and try to deploy the stack I'm seeing the following error `The specified KMS key does not exist or is not allowed to be used with Arn 'arn:aws:logs:us-east-1:<account_id>:log-group:/aws/vendedlogs/states/MyLogGroup`. If I remove the encryption context it deploys successfully, not sure what I'm doing wrong here?:

```typescript
logGroupKey.addToResourcePolicy(new cdk.aws_iam.PolicyStatement({
      resources: ['*'],
      actions: ['kms:Encrypt*', 'kms:Decrypt*', 'kms:ReEncrypt*', 'kms:GenerateDataKey*', 'kms:Describe*'],
      principals: [new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`)],
      conditions: {
        ArnEquals: {
          'kms:EncryptionContext:aws:logs:arn': cdk.Stack.of(this).formatArn({
            service: 'logs',
            resource: 'log-group',
            sep: ':',
            resourceName: 'MyLogGroup',
          }),
        },
      },
    }));`
```

Reply:
I've removed this line in the [latest commit](https://github.com/aws/aws-cdk/pull/30959/commits/d668d9b6841e05e59213aac069d803b422565cce), the log group key will use encryption context and the regional CWL service principal.